ui/job-view/redux/stores/selectedJob.js

```diff
@@ -0,0 +1,231 @@
+import { getUrlParam, setUrlParam } from '../../../helpers/location';
+import JobModel from '../../../models/job';
+import PushModel from '../../../models/push';
+import { getJobsUrl } from '../../../helpers/url';
+
+const SELECT_JOB = 'SELECT_JOB';
+const SELECT_JOB_FROM_QUERY_STRING = 'SELECT_JOB_FROM_QUERY_STRING';
+const CLEAR_JOB = 'CLEAR_JOB';
+const UPDATE_JOB_DETAILS = 'UPDATE_JOB_DETAILS';
+
+export const setSelectedJob = (job, updateDetails = true) => ({
+  type: SELECT_JOB,
+  job,
+  updateDetails,
+});
+
+export const setSelectedJobFromQueryString = (notify, jobMap) => ({
+  type: SELECT_JOB_FROM_QUERY_STRING,
+  notify,
+  jobMap,
+});
+
+export const clearSelectedJob = countPinnedJobs => ({
+  type: CLEAR_JOB,
+  countPinnedJobs,
+});
+
+export const updateJobDetails = job => ({
+  type: UPDATE_JOB_DETAILS,
+  job,
+  meta: {
+    debounce: 'nextJob',
+  },
+});
+
+const doUpdateJobDetails = job => {
+  const jobId = job ? job.id : null;
+
+  setUrlParam('selectedJob', jobId);
+  return { selectedJob: job };
+};
+
+export const doSelectJob = (job, updateDetails) => {
+  const selected = findSelectedInstance();
+
+  if (selected) selected.setSelected(false);
+
+  const newSelectedElement = findJobInstance(job.id, true);
+
+  if (newSelectedElement) {
+    newSelectedElement.setSelected(true);
+  } else {
+    const group = findGroupInstance(job);
+    if (group) {
+      group.setExpanded(true);
+    }
+
+    // If the job is in a group count, then the job element won't exist, but
+    // its group will.  We can try scrolling to that.
+    const groupEl = findGroupElement(job);
+    if (groupEl) {
+      scrollToElement(groupEl);
+    }
+  }
+  if (updateDetails) {
+    return doUpdateJobDetails(job);
+  }
+  return { selectedJob: job };
+};
+
+export const doClearSelectedJob = countPinnedJobs => {
+  if (!countPinnedJobs) {
+    const selected = findSelectedInstance();
+    if (selected) selected.setSelected(false);
+    setUrlParam('selectedJob', null);
+    return { selectedJob: null };
+  }
+  return {};
+};
+
+/**
+ * If the URL has a query string param of ``selectedJob`` then select
+ * that job on load.
+ *
+ * If that job isn't in any of the loaded pushes, then throw
+ * an error and provide a link to load it with the right push.
+ */
+const doSetSelectedJobFromQueryString = (notify, jobMap) => {
+  const repoName = getUrlParam('repo');
+  const selectedJobIdStr = getUrlParam('selectedJob');
+  const selectedJobId = parseInt(selectedJobIdStr, 10);
+
+  if (selectedJobIdStr) {
+    const selectedJob = jobMap[selectedJobIdStr];
+
+    // select the job in question
+    if (selectedJob) {
+      return doSelectJob(selectedJob);
+    }
+    setUrlParam('selectedJob');
+    // If the ``selectedJob`` was not mapped, then we need to notify
+    // the user it's not in the range of the current result set list.
+    JobModel.get(repoName, selectedJobId)
+      .then(job => {
+        PushModel.get(job.push_id).then(async resp => {
```

Comment:
This method should probably be changed to use the `getData` wrapper at some point; there's only a few places where its being used.

Reply:
Yeah, sounds good.  I'll do it in a follow-up PR, if that's cool.  :)